### PR TITLE
add support for no-index, requirements files, and repo paths

### PR DESCRIPTION
**Pull request by georgeliaw**
Needed to update the `pex_binary` rules to keep the clutter down in the BUILD files and make them easier to maintain. Figured I should open a PR in case you would like to pull this into master.

### CODE REVIEW

File: pex/pex_rules.bzl [hunk 1]
```diff
@@ -82,6 +83,16 @@ def _collect_transitive_reqs(ctx):
   return transitive_reqs
 
 
+def _collect_repos(ctx):
+  repos = {}
+  for dep in ctx.attr.deps:
+    if hasattr(dep.py, "repos"):
+      repos += dep.py.repos
+  for file in egg_file_types.filter(ctx.files.repos):
```
Comment (benley) [inline pex/pex_rules.bzl]: Would it work to allow source distributions in `repos` along with eggs and wheels? That could be very convenient at times.

Comment (benley) [inline pex/pex_rules.bzl]: Also technically one can specify https URLs pointing at a package repository with `--repos`.  It's not a great practice with regard to Bazel, but I know people will want to do that... Thoughts?

Comment (georgeliaw) [inline pex/pex_rules.bzl]: @benley added support for source distributions as well as cache disabling.

Have some mixed feelings about supporting URLs for the same reason that it's not exactly great practice. I'll think on the issue a bit this week.

File: pex/BUILD [hunk 2]
```diff
@@ -22,7 +22,7 @@ genrule(
         # Workaround really long shebang lines breaking on linux:
         # Use a /tmp path, but keep the actual venv inside the bazel outdir.
         # Avoids having to worry about cleanup, even if sandboxing is off.
-        TMPF=$$(mktemp)
+        TMPF=$$(mktemp 2>/dev/null || mktemp -t tmp)
```
Comment (benley) [inline pex/BUILD]: Wow, I wonder how this ever worked on MacOS - did mktemp's behaviour change in a recent macos release?

Comment (georgeliaw) [inline pex/BUILD]: @benley yeah this was a weird one, it appears the behavior was fixed to match gnu in OS X 10.11, but older versions of OS X require the `-t` flag. Only encountered it due to a coworker running into this issue.

https://unix.stackexchange.com/questions/30091/fix-or-alternative-for-mktemp-in-os-x

**Comment by georgeliaw** [comment]
@benley gave the URL `--repo` some thought. I think there's more complexity to that task than would fit into the scope of this PR. For the time being, users should be able to pull packages in via `http_file` in their workspace (possibly wrap it up into a filegroup) and load it into either the `eggs` or `repos` params. Have not tried that out myself, but this way they would be able to benefit from checksum enforcement.

Anything nicer would probably require writing new workspace rules for like `pypi_server` to set the url and `pypi_package` to set package name/checksum. Not sure that work is worth doing at the moment if we can use the workaround above.

For the time being, we can consider the PR ready to merge unless there are any changes you would like me to make.

**Comment by georgeliaw** [comment]
@benley I just recalled that requirements files themselves can be used to point to other repo urls, so technically this is already supported. Again, obviously not the greatest practice when it comes to Bazel, but the use-case is available. 
https://pip.pypa.io/en/stable/reference/pip_install/#requirements-file-format

Pip since version 8.0 also supports hash checking, which we could try enforcing to improve the build determinism:
https://pip.pypa.io/en/stable/reference/pip_install/#hash-checking-mode

This technically means we wouldn't have to go down the workspace rule route.

What are your thoughts on this?

**Comment by benley** [comment]
Oh cool, that is a good point.  I completely forgot about this PR, sorry about that!

**Pull request merged**